flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableImpl.java

```diff
@@ -134,6 +143,12 @@ public TaskSlotTableImpl(
 		slotActions = null;
 		state = State.CREATED;
 		closingFuture = new CompletableFuture<>();
+
+		asyncExecutor = new ThreadPoolExecutor(
+			0,
+			numberSlots,
+			60L, TimeUnit.SECONDS,
+			new SynchronousQueue<>());
```

Comment:
Whenever you use an executor that is not provided via some factory method in `java.util.concurrent.Executors` I would recommend writing down some tests for it to ensure it behaves exactly the way you expect it to.
This right here is probably fine since we know the upper limit of possible tasks to be queued (== 1 task per slot), but a test would also be good for documentation purposes.

Reply:
What kind of test do you mean? I just want to use it as a cached executor. I want it to allocate a new thread if there is no idle one for each slot closing.

Reply:
We decided to switch to `TaskManagerServices#ioExecutor` in #12980.